File: Blind 75/16 - Intervals/57-insert-intervals.py
'''
You are given an array of non-overlapping intervals intervals where intervals[i] = [start_i, end_i] represent the start and the end of the ith interval and intervals is sorted in ascending order by start_i. You are also given an interval newInterval = [start, end] that represents the start and end of another interval.

Insert newInterval into intervals such that intervals is still sorted in ascending order by start_i and intervals still does not have any overlapping intervals (merge overlapping intervals if necessary).

Return intervals after the insertion.
'''
from time import time
from typing import List


class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        res = []
        for i in range(len(intervals)):
            if newInterval[1] < intervals[i][0]:
                # New interval is less than upcoming interval
                res.append(newInterval)
                return res + intervals[i:]
            elif newInterval[0] > intervals[i][1]:
                # New interval is greater than upcoming interval
                res.append(intervals[i])
            else:
                # Otherwise, interval overlaps
                newInterval = [min(newInterval[0], intervals[i][0]),
                               max(newInterval[1], intervals[i][1])]

        res.append(newInterval)
        return res

    def reference(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        res = []

        for i in range(len(intervals)):
            if newInterval[1] < intervals[i][0]:
                res.append(newInterval)
                return res + intervals[i:]
            elif newInterval[0] > intervals[i][1]:
                res.append(intervals[i])
            else:
                newInterval = [
                    min(newInterval[0], intervals[i][0]),
                    max(newInterval[1], intervals[i][1]),
                ]
        res.append(newInterval)
        return res

    def quantify(self, test_cases, runs=100000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.insert(case[0], case[1]))
                else:
                    self.insert(case[0], case[1])
        print(f'Runtime for our solution: {time() - sol_start}')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case[0], case[1]))
                else:
                    self.reference(case[0], case[1])
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        ([[1, 3], [6, 9]], [2, 5]),
        ([[1, 2], [3, 5], [6, 7], [8, 10], [12, 16]], [4, 8])
    ]
    test.quantify(test_cases)

File: Blind 75/16 - Intervals/JavaScript/57-insert-intervals.js
const { performance } = require('perf_hooks');

class Solution {
  method(intervals, newInterval) {
    let res = [];
    for (let i = 0; i < intervals.length; i++) {
      if (newInterval[1] < intervals[i][0]) {
        res.push(newInterval);
        res.push(intervals.slice(i));
        return res;
      } else if (newInterval[0] > intervals[i][1]) {
        res.push(intervals[i]);
      } else {
        newInterval = [
          Math.min(newInterval[0], intervals[i][0]),
          Math.max(newInterval[1], intervals[i][1]),
        ];
      }
    }

    res.push(newInterval);
    return intervals;
  }

  reference(intervals, newInterval) {
    const { beforeIndex, before } = this.getBefore(intervals, newInterval);
    const afterIndex = this.mergeIntervals(intervals, newInterval, beforeIndex);
    const after = intervals.slice(afterIndex);

    return [...before, newInterval, ...after];
  }

  getBefore = (intervals, newInterval, index = 0, before = []) => {
    const hasGap = ([prevStart, prevEnd], [currStart, currEnd]) =>
      prevEnd < currStart;

    while (index < intervals.length && hasGap(intervals[index], newInterval)) {
      const current = intervals[index];

      before.push(current);
      index++;
    }

    return { beforeIndex: index, before };
  };

  mergeIntervals = (intervals, newInterval, index) => {
    const hasOverlap = ([prevStart, prevEnd], [currStart, currEnd]) =>
      currStart <= prevEnd;

    while (
      index < intervals.length &&
      hasOverlap(newInterval, intervals[index])
    ) {
      const current = intervals[index];

      newInterval[0] = Math.min(newInterval[0], current[0]);
      newInterval[1] = Math.max(newInterval[1], current[1]);
      index++;
    }

    return index;
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(...input));
        else this.method(...input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(...input));
        else this.reference(...input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [
    [
      [1, 3],
      [6, 9],
    ],
    [2, 5],
  ],
  [
    [
      [1, 2],
      [3, 5],
      [6, 7],
      [8, 10],
      [12, 16],
    ],
    [4, 8],
  ],
];
test.quantify(testCases);
